SignalMessaging/attachments/EditContactShareNameViewController.swift

```diff
@@ -0,0 +1,308 @@
+    }
+
+    public func value() -> String {
+        guard let value = valueView.text else {
+            return ""
+        }
+        return value
+    }
+}
+
+// MARK: -
+
+@objc
+public protocol EditContactShareNameViewControllerDelegate: class {
+    func editContactShareNameView(_ editContactShareNameView: EditContactShareNameViewController, didEditContactShare contactShare: OWSContact)
+}
+
+// MARK: -
+
+@objc
+public class EditContactShareNameViewController: OWSViewController, ContactNameFieldViewDelegate {
+    weak var delegate: EditContactShareNameViewControllerDelegate?
+
+    let contactShare: OWSContact
+
+    var namePrefixView: ContactNameFieldView!
+    var givenNameView: ContactNameFieldView!
+    var middleNameView: ContactNameFieldView!
+    var familyNameView: ContactNameFieldView!
+    var nameSuffixView: ContactNameFieldView!
+
+    var fieldViews = [ContactNameFieldView]()
+
+    // MARK: Initializers
+
+    @available(*, unavailable, message:"use other constructor instead.")
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("unimplemented")
+    }
+
+    @objc
+    required public init(contactShare: OWSContact, delegate: EditContactShareNameViewControllerDelegate) {
+        self.contactShare = contactShare
+        self.delegate = delegate
+
+        super.init(nibName: nil, bundle: nil)
+
+        buildFields()
+    }
+
+    func buildFields() {
+        namePrefixView = ContactNameFieldView(name: NSLocalizedString("CONTACT_FIELD_NAME_PREFIX", comment: "Label for the 'name prefix' field of a contact."),
+                                              value: contactShare.namePrefix, delegate: self)
+        givenNameView = ContactNameFieldView(name: NSLocalizedString("CONTACT_FIELD_GIVEN_NAME", comment: "Label for the 'given name' field of a contact."),
+                                             value: contactShare.givenName, delegate: self)
+        middleNameView = ContactNameFieldView(name: NSLocalizedString("CONTACT_FIELD_MIDDLE_NAME", comment: "Label for the 'middle name' field of a contact."),
+                                              value: contactShare.middleName, delegate: self)
+        familyNameView = ContactNameFieldView(name: NSLocalizedString("CONTACT_FIELD_FAMILY_NAME", comment: "Label for the 'family name' field of a contact."),
+                                              value: contactShare.familyName, delegate: self)
+        nameSuffixView = ContactNameFieldView(name: NSLocalizedString("CONTACT_FIELD_NAME_SUFFIX", comment: "Label for the 'name suffix' field of a contact."),
+                                              value: contactShare.nameSuffix, delegate: self)
+        fieldViews = [
+            namePrefixView ,
+            givenNameView ,
+            middleNameView ,
+            familyNameView ,
+            nameSuffixView
+        ]
+    }
+
+    override public var canBecomeFirstResponder: Bool {
+        return true
```

Comment:
again, why/when do we become first responder?

Reply:
See above.